### PR TITLE
Windows: compile xz with CMake

### PR DESCRIPTION
**Pull request by nulano**
Changes proposed in this pull request:

 * Compile liblzma with CMake instead of VS project file, for https://github.com/python-pillow/Pillow/pull/6883#issuecomment-1379657783:
   > Updated windows/INSTALL-MSVC.txt to mention that CMake-based build is now the preferred method with Visual Studio. The project files will probably be removed after 5.4.x releases.

   This might also help with https://github.com/python-pillow/Pillow/issues/6679#issuecomment-1296080510
 * Use Ninja instead of NMake to build dependencies by default. Unlike NMake, Ninja can build several files in parallel. This makes very little difference on GHA which only provides 2 CPU cores, but is almost 2x faster on my computer.
 * Rewrite argument parsing in `winbuild\build_prepare.py` using `argparse` for better maintainability.

### CODE REVIEW

File: winbuild/build_prepare.py [hunk 1]
```diff
@@ -593,92 +576,138 @@ def build_pillow():
...
+        metavar="PILLOW_BUILD",
+        default=os.environ.get("PILLOW_BUILD", os.path.join(winbuild_dir, "build")),
+        help="build directory (default: 'winbuild\\build')",
+    )
+    parser.add_argument(
+        "--depends",
+        dest="depends_dir",
+        metavar="PILLOW_DEPS",
+        default=os.environ.get("PILLOW_DEPS", os.path.join(winbuild_dir, "depends")),
+        help="directory used to store cached dependencies (default: 'winbuild\\depends')",  # noqa: E501
```
Comment (hugovk) [inline winbuild/build_prepare.py:606]:
```suggestion
        help="directory used to store cached dependencies "
        "(default: 'winbuild\\depends')",
```

File: winbuild/build_prepare.py [hunk 2]
```diff
@@ -593,92 +576,138 @@ def build_pillow():
...
+            ),
+        ),
+        help="build architecture (default: same as host python)",
+    )
+    parser.add_argument(
+        "--python",
+        dest="python_dir",
+        metavar="PYTHON",
+        default=os.environ.get("PYTHON"),
+        help="Python install directory (default: use host python)",
```
Comment (hugovk) [inline winbuild/build_prepare.py:626]:
```suggestion
        help="Python install directory (default: use host Python)",
```

File: winbuild/build_prepare.py [hunk 3]
```diff
@@ -593,92 +576,138 @@ def build_pillow():
...
+    )
+    parser.add_argument(
+        "--executable",
+        dest="python_exe",
+        metavar="EXECUTABLE",
+        default=os.environ.get("EXECUTABLE", "python.exe"),
+        help="Python executable (default: use host python)",
```
Comment (hugovk) [inline winbuild/build_prepare.py:633]:
```suggestion
        help="Python executable (default: use host Python)",
```

File: winbuild/build_prepare.py [hunk 4]
```diff
@@ -593,92 +576,138 @@ def build_pillow():
...
+        help="skip LGPL-licensed optional dependency FriBiDi",
+    )
+    args = parser.parse_args()
 
-    if python_dir is None:
-        python_dir = os.path.dirname(os.path.realpath(sys.executable))
-        python_exe = os.path.basename(sys.executable)
-    print("Target Python:", os.path.join(python_dir, python_exe))
+    arch_prefs = architectures[args.architecture]
+    print("Target Architecture:", args.architecture)
```
Comment (hugovk) [inline winbuild/build_prepare.py:657]:
```suggestion
    print("Target architecture:", args.architecture)
```

**Comment by hugovk** [comment]
> * Use Ninja instead of NMake to build dependencies by default. Unlike NMake, Ninja can build several files in parallel. This makes very little difference on GHA which only provides 2 CPU cores, but is much faster on my computer.

Sounds good! What sort of speedup do you see?

**Comment by nulano** [comment]
> Sounds good! What sort of speedup do you see?

Almost 2x on my 6 core 12 thread CPU.

With ``--nmake``:
```
Days              : 0
Hours             : 0
Minutes           : 6
Seconds           : 44
Milliseconds      : 37
Ticks             : 4040372902
TotalDays         : 0.00467635752546296
TotalHours        : 0.112232580611111
TotalMinutes      : 6.73395483666667
TotalSeconds      : 404.0372902
TotalMilliseconds : 404037.2902
```

Without ``--nmake`` (i.e. with Ninja):
```
Days              : 0
Hours             : 0
Minutes           : 3
Seconds           : 39
Milliseconds      : 489
Ticks             : 2194899729
TotalDays         : 0.00254039320486111
TotalHours        : 0.0609694369166667
TotalMinutes      : 3.658166215
TotalSeconds      : 219.4899729
TotalMilliseconds : 219489.9729
```

**Review by hugovk — APPROVED** [review]

File: winbuild/build_prepare.py [hunk 5]
```diff
@@ -55,24 +56,28 @@ def cmd_nmake(makefile=None, target="", params=None):
...
+                "-DCMAKE_C_COMPILER=cl.exe",  # for Ninja
+                "-DCMAKE_CXX_COMPILER=cl.exe",  # for Ninja
+                "-DCMAKE_C_FLAGS=-nologo",
+                "-DCMAKE_CXX_FLAGS=-nologo",
+                *params,
+                '-G "{cmake_generator}"',
+                ".",
+            ]
+        ),
+        f"{{cmake}} --build . --clean-first --parallel --target {target}",
```
Comment (nulano) [inline winbuild/build_prepare.py:79]: I originally updated the documentation in this PR to require CMake 3.13+ as that is what xz's CMake files specify.

I missed that the `cmake --build` command requires CMake 3.15+ when there are multiple targets, I've now updated the PR.

The previous CMake requirement was 3.12, which is what is included with VS2017 and was the oldest I tested when writing the documentation in #4495.

VS2019 includes CMake 3.20, but I've found that 3.15 is sufficient for this PR.

**Pull request merged**